### DISCUSSION TITLE
A function to go back/forward like in browsers

### DISCUSSION THREAD

**Discussion by akippnn** [ppy/osu#17577] [Ideas] [answered]
A "go back"/"go forward" function found in browsers

I believe it is essential because the laser client serves web content, so navigating from A > B > C > D and back to C means I have to navigate A > B > C again, and the page itself isn't cached so I have to scroll to the position I was in.

How to reproduce:
1. Go to a player profile
2. Find and go to a beatmap from the player profile
3. Select a player from beatmap leaderboards
4. Open another beatmap from the current player profile
5. If I want to go back to the second player that I found from the leaderboards, I need to do step 1-3 again.

The feature may already exist as pressing ESC/M4/M5 but I think it should be defined as "going up a level" internally. Especially since it's confusing whether or not the menus are or would be nested, whether the player should expect if they go back from the previous navigated menu or not.

This would also be helpful if I were navigating someone's profile and I want to open chat, then go back to the said profile. The client should be aware of what I opened last.

The function can be used in 3 ways:
- Buttons in the toolbar
  - A "go back" button
  - A "go forward" button
- Respond from mouse buttons (M4/M5)
- Respond from keyboard shortcuts
  - Default for back/forward would be alt+left/alt+right

Other changes that have to be made after this is implemented:
- Define ESC as an "escape" key in Key Configuration settings.
- Treat content shown on the screen as layers to make the interface more intuitive, this way the player knows what to expect when using the Escape function or the Back function.
- Do not rely on nesting (especially in web content such as beatmap listing).

Frankly I am not quite sure how this would pan out since I didn't give it enough thought (it only popped in my head an hour from now), but I think osu's interface needs to be reimagined a little bit. I would love an open discussion about this with you guys, especially with the people responsible for the game's design.

**Comment by frenzibyte** [comment, marked as answer]
This is essentially a duplicate of #5063, <kbd>ESC</kbd> is known to be a back key in the game code, so no changes are required to support going back between overlays. Nesting doesn't sound like it would be harmful in any way either, as it essentially just opens another overlay.

**Reply by akippnn** [reply]
Sorry for the duplicate!

Going forwards, I will try to participate in that issue if I can. Thank you.
